### ISSUE TITLE
Build Validation Layers with Android CMake Plugin

### ISSUE THREAD

**Issue by byumjin** [KhronosGroup/Vulkan-ValidationLayers#984]
Hello,

I'm trying to Build Validation Layers with Android CMake Plugin in my android studio project.
Currently, it gives error like below:
![캡처_2019_06_16_08_08_19_280](https://user-images.githubusercontent.com/22266515/59565929-076c7100-900e-11e9-9ffa-118a510dd1b7.png)

If I execute cmake with the command line, the prompt shows the error like below:
![캡처_2019_06_16_08_10_43_68](https://user-images.githubusercontent.com/22266515/59565957-41d60e00-900e-11e9-9c80-024c0117829e.png)

I guess, somehow, cmake can't generate "generator Android".
I've tried the build with different NDK versions. But the result is the same.

Any suggestion?

My working environment is like this:

- OS: Windows 10 10.0
- Android Studio: 3.4.1
- Build #AI-183.6156.11.34.5522156, built on May 1, 2019
- JRE: 1.8.0_152-release-1343-b01 amd64
- JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
- Target SDK Version: 24
- Compile SDK Version: 24
- NDK Version: 20
- Gradle Version: 5.1.1

**Comment by mikew-lunarg** [comment]
Looking at the CMake output in the second image, it appears the generator name might be "Android Gradle - Ninja" rather than just "Android"

**Comment by mikew-lunarg** [comment]
Also, in that same image, the CMake invocation appears to be missing quotes around the generator name.

**Comment by mark-lunarg** [comment]
@cnorthrop, any ideas?

**Comment by byumjin** [comment]
Thank you, after I added the quotes manually, it could pass the issue.
But CMake errors are still going on. I'm going to try to find out what is the reason of theses. 
![Capture_2019_06_17_10_59_52_839](https://user-images.githubusercontent.com/22266515/59626926-eaa96980-90f1-11e9-8cc7-a138606edb0a.png)

**Comment by mikew-lunarg** [comment]
I'm not expert on Android Studio and its CMake plugin.   From where did you get the ValidationLayer source code?  What is its version/commit ID?
FWIW the CMake version shown in your output (3.6.4...) is below the recommended 3.10.2, don't know if that could be contributing.

**Comment by byumjin** [comment]
I got the ValidationLayer codes from this repository ( tests: Fix broken tests related to query states: 8df3ad8df3 )
I tried cmake 3.10.2 but it gave the same result.

**Comment by cnorthrop** [comment]
We've got instructions for how to build with Android Studio ([here](https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/master/build-android/cmake/README.md)) but we don't actively test it with CI.

@ggfan Any insights into what could cause this?

**Comment by byumjin** [comment]
@cnorthrop Yes, I already followed the guidance.

By the way, if I removed add_library parts for 'VkLayer_khronos_validation' in CMakeList.txt (line 62 to 96)
The first error is gone. 

And for the second and the third error, it seems like it needs ${SRC_DIR}/include.
But, the folder doesn't exist in my local disk.
But, weird thing is.. other layers also need the include folder for 'target_include_directories' but doesn't complain about this.

I will share my folder hierarchy in here

![Capture_2019_06_17_11_52_04_240](https://user-images.githubusercontent.com/22266515/59628910-8a68f680-90f6-11e9-85fa-5c9ed026626c.png)

![Capture_2019_06_17_11_52_10_34](https://user-images.githubusercontent.com/22266515/59628912-8ccb5080-90f6-11e9-99a2-2b05beee751d.png)

**Comment by ggfan** [comment]
Inside NDK there is an old version of the code for the validation layers, could you try to build it from there to see if that works on your system? %NDK-dir%\source\third_party\vulkan\src.

that would know whether something missed from the latest code inside this repo

**Comment by byumjin** [comment]
Although, I set BUILD_IN_NDK to be ON always (in CMakelist.txt), the result was not changed.

**Comment by byumjin** [comment]
I think I found the reasons.

There are parenthesis typos.

In Line 158, there is a missing parenthesis.
In Line 168, there is an extra parenthesis.

Could you guys check this?

**Comment by byumjin** [comment]
Currently, I got this error when I tried to build Validation Layers.
![Capture_2019_06_17_13_58_23_496](https://user-images.githubusercontent.com/22266515/59636346-22231080-9108-11e9-9feb-7e570ae27484.png)

Any ideas?

**Comment by ggfan** [comment]
check your NDK version: AHardwareBuffer is Android specific thing.
 - make sure used NDK has it ( inside header file directory )
 - the android version NDK began to expose it
 - your build.gradle's minimum api level

somewhere along the line, you would be able to fix it.

**Comment by cnorthrop** [comment]
@mark-lunarg I think the CMakelist.txt changes came in 1f2ba2625 and 60338829e.  @byumjin, can you submit a patch that fixes those?

Regarding `AHardwareBuffer_describe`, those entrypoints were added in Android O (API level 26).  Compiling for API level 24 (Nougat) may no longer be supported.  @daveh-lunarg, can you confirm?  @byumjin, can you target API level 26 (or better, 28) instead?

If we are going to support compiling with Android Studio, we should get this hooked up to CI.  We're currently using ndk-build everywhere for validation.

**Comment by byumjin** [comment]
@cnorthrop for fixing CMakelist.txt, I still have a issue with 'add_library cannot create imported target "SPIRV-Tools-opt-prebuilt"'. I think I can't remove the part for 'VkLayer_khronos_validation', but it makes issue for me. Do you want me to fix only parenthesis issue?

Regarding 'AHardwareBuffer_describe' issue, i will try other NDK and SDK, and let you guys know.

**Comment by byumjin** [comment]
So, if I use SDK version 24, it gives error at stage 10.

![Capture_2019_06_17_15_57_30_517](https://user-images.githubusercontent.com/22266515/59642325-5eab3800-9119-11e9-9901-1395de7e88df.png)

But, if I use SDK version higher that 26, it passes the stage10, but it is stuck at stage 11.
And report a bunch of  undefined errors related to 'Optimizer'.

![Capture_2019_06_17_15_59_59_106](https://user-images.githubusercontent.com/22266515/59642328-62d75580-9119-11e9-826f-7075e58fc774.png)

I'm currently using NDK 20 for Environment Values, Android Studio's project

**Comment by ggfan** [comment]
could you point me to the doc you are mentioning about stage 10 stage 11? thanks

**Comment by byumjin** [comment]
I'm not sure which doc executes the commands.
But, the command arguments like this
'cmake --build F:\validationlayers\Vulkan-ValidationLayers\build-android\cmake\layerlib\.externalNativeBuild\cmake\debug\arm64-v8a --target VkLayer_core_validation'

**Comment by byumjin** [comment]
It seems like the above errors are coming from referring to wrong include header files.
If I'm correct, CMake should refer to '/validationlayers/Vulkan-ValidationLayers/layers' to use AHardwareBuffer_describe()

But, according to the logs, it seems like CMake looked for the include folder in order below

-IF:/validationlayers/Vulkan-ValidationLayers/include
-IF:/validationlayers/Vulkan-ValidationLayers/build-android/generated/include
-IF:/validationlayers/Vulkan-ValidationLayers/layers
-IF:/validationlayers/Vulkan-ValidationLayers/loader
-IF:/validationlayers/Vulkan-ValidationLayers/build-android/third_party/shaderc/third_party/glslang
-IF:/validationlayers/Vulkan-ValidationLayers/build-android/third_party/shaderc/third_party/spirv-tools/include

I'm not used to use CMake before, but what I understood is, if there is a folder in the list (from very first one) , CMake does not seem to include the rest of them.

Currently, I have '/validationlayers/Vulkan-ValidationLayers/build-android/generated/include' so I would like to make CMake traverse '/validationlayers/Vulkan-ValidationLayers/layers' first. So, I changed target_include_directories for VkLayer_core_validation.
![Capture_2019_06_18_08_41_27_515](https://user-images.githubusercontent.com/22266515/59698665-0752a980-91a5-11e9-81b3-eafcc262bea3.png)

But, the result order is not changed, even if I removed the CMake cache from the CMakelist.txt.

**Comment by cnorthrop** [comment]
@byumjin While we sort out cmake issues, are you able to build layers from the command line, just to get you unblocked?  Android steps are [here](https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/master/BUILD.md#building-on-android).

**Comment by byumjin** [comment]
with the command line
'cd build-android'
'update_external_sources_android.bat'
'android-generate.bat'
'ndk-build'

It seemed fine, I couldn't find any issue.

**Comment by byumjin** [comment]
After I wiped my local Vulkan Validation Layer repo out and re-cloned and built it,
I encountered new issue 'use of undeclared identifier 'VK_DRIVER_ID_GOOGLE_PASTEL_KHR' when I tried to build layerlib

![Capture_2019_06_18_11_26_05_983](https://user-images.githubusercontent.com/22266515/59709929-bcdd2700-91bc-11e9-9207-13167ccf7a17.png)

- OS: Windows 10 10.0
- Android Studio: 3.4.1
- Build #AI-183.6156.11.34.5522156, built on May 1, 2019
- JRE: 1.8.0_152-release-1343-b01 amd64
- JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
- Target SDK Version: 26
- Compile SDK Version: 26
- NDK Version: 20
- Gradle Version: 5.1.1

**Comment by Andreyogld3d** [comment]
@byumjin, BTW I think this is not critical bug, we can build from command line, it works very well.

**Comment by byumjin** [comment]
@Andreyogld3d could you run command line below?
'cmake --build F:\validationlayers\Vulkan-ValidationLayers\build-android\cmake\layerlib\.externalNativeBuild\cmake\debug\arm64-v8a --target VkLayer_core_validation'

 I tried it but it showed the same result...

**Comment by ggfan** [comment]
this is hard part: directly invoking from CMake on command line may be too much ( there are a few factors may not worth the time to communicate ), so I would stay with gradle file: might be more fruitful ( even personally I do not feel pride in saying so :-( ).

you already have an app, include this project as a subproject and keeps on debug from there. I would try to see what you saw probably some later time after I put my current work in-shape: to get to Android build needs to get many dependencies done in this repo... 
If you do not have app, you could get some idea from this one: https://github.com/googlesamples/android-vulkan-tutorials/tree/master/tutorial03_traceable_layers
(but you could not use exactly as-is: that one using source inside NDK )

**Comment by byumjin** [comment]
@ggfan Thank you, i will check the link. Also, I will share my steps to build laylib here

**Comment by Andreyogld3d** [comment]
@byumjin 

> cmake --build F:\validationlayers\Vulkan-ValidationLayers\build-android\cmake\layerlib.externalNativeBuild\cmake\debug\arm64-v8a --target VkLayer_core_validation

Doesn't work, because i have no directory .externalNativeBuild\cmake\debug\arm64-v8a.

**Comment by byumjin** [comment]
I tried to build this a lot of times (with different NDK and SDK versions), unfortunately, I couldn't make it. It always stuck with "undefined reference to `AHardwareBuffer_describe'" error. It seems like AHardwareBuffer_describe() is supported by recent NDK and SDK version. So, I also tried NDK20 version but it keeps complain about it. 

My steps for building Validation layer, which is traceable, is like below:

OS: Windows 10 10.0
Android Studio: 3.4.1
Build #AI-183.6156.11.34.5522156, built on May 1, 2019
JRE: 1.8.0_152-release-1343-b01 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
NDK Version: 20
Gradle Version: 5.1.1

**1. Clone the latest Vulkan Validation Repo** form [here ](https://github.com/KhronosGroup/Vulkan-ValidationLayers) to next your android project. In my case, I'm using [Oculus Mobile Sdk](https://developer.oculus.com/downloads/package/oculus-mobile-sdk/).

**2. Build Dependent Repositories with Known-Good Revisions (I'm using Windows10 x64)**

> cd Vulkan-ValidationLayers
mkdir build
cd build
..\scripts\update_deps.py --arch x64
cmake -A x64 -C helper.cmake ..
cmake --build 
>

**3. Build Android stuffs**

Set your Environment variables for Android
> set ANDROID_SDK_HOME=%LOCALAPPDATA%\Android\sdk
set ANDROID_NDK_HOME=%LOCALAPPDATA%\Android\sdk\ndk-bundle
set PATH=%LOCALAPPDATA%\Android\sdk\ndk-bundle;%PATH%
>

And build it

> cd build-android
update_external_sources_android.bat
android-generate.bat
ndk-build (This is optional I think)
>

**4. Open your project and link the `layerlib` project**

In my case, my main porject was VrCubeWorld_Vulkan`.
I needed to convert Task `<<` operator to `doLast{}` for my gradle (VrApp.gradle).
add `':layerLib'` to `setting.gradle` in VrCubeWorld_Vulkan like
>include ':', 'VrSamples:VrCubeWorld_Vulkan:Projects:Android', ':layerLib'
>
add project to `setting.gradle` in VrCubeWorld_Vulkan like
For example,
>project(':layerLib').projectDir = new File('../../../../../Vulkan-ValidationLayers/build-android/cmake/layerlib')
>

Now, your app should recognize the new added project (layerLib)

Go to Project Structure -> Module
>Set your project's Compile Sdk Version to 26 in Properties tap
Set your project's Target Sdk Version to 26 in Default Config tap
Set your project's Min Sdk Version to 26 in Default Config tap
>

>Set your layeLib's Compile Sdk Version to 26 in Properties tap
Set your layeLib's Target Sdk Version to 26 in Default Config tap
Set your layeLib's Min Sdk Version to 26 in Default Config tap
>

You may also need to fix `'-DANDROID_PLATFORM=android-24'` to `'-DANDROID_PLATFORM=android-26'` at cmake.arguments from `build.gradle` file in `layerLib` project.

Now, if you build the project. You may encounter this errors.
![Capture_2019_06_19_17_55_02_75](https://user-images.githubusercontent.com/22266515/59810753-8b975080-92bb-11e9-9696-ec33d5427ccc.png)

**5. Fix CMakeLists.txt**

_From this step, I might be wrong._
To avoid the `another target with the same name already exists` issue, I removed line 62 ~ 96 because I just want to use old Validation layers.
You will still encounter missing `VkLayer_unique_objects` errors.
Actually, there are bracket typos in the file.

at `add_library(VkLayer_threading SHARED`
add ')' to `${COMMON_DIR}/include/chassis.cpp`

>
`${COMMON_DIR}/include/chassis.cpp -> ${COMMON_DIR}/include/chassis.cpp)`
>
And, at `add_library(VkLayer_unique_objects SHARED`
 remove ')' from `${COMMON_DIR}/include/layer_chassis_dispatch.cpp)`

>
`${COMMON_DIR}/include/layer_chassis_dispatch.cpp) -> ${COMMON_DIR}/include/layer_chassis_dispatch.cpp`
>
 
Now, if you build the project. You may encounter missing `vulkan/vk_layer.h` error, because there is no header folder in our Vulkan validation folder. So, I copied `include` folder from `Vulkan-ValidationLayers\build\Vulkan-Headers\build\install\include`. **I'm not sure** this is the right solution.

If you build it again, you will face `undefined reference to `AHardwareBuffer_describe'` error.
What I don't understand is, even if I set my min SDK version as 26 and NDK version as 20, it keeps complains about it.

Any suggestions will be helpful.
Thank you.

**Comment by byumjin** [comment]
So, I figure out that `layers/core_validation.cpp` includes `android_ndk_type.h`, which includes `android/hardware_buffer.h`.
In `android_ndk_type.h`, I've checked that my __ANDROID_API__ is 26.
And, In `android/hardware_buffer.h`, which is from `android-ndk-r20/toolchains/llvm/prebuilt/windows-x86_64/sysroot/usr/include/android`, I've also checked that line 335 `#if __ANDROID_API__ >= 26` passed.

![Capture_2019_06_20_12_42_52_275](https://user-images.githubusercontent.com/22266515/59876863-812b9400-9359-11e9-8586-9a2dbe0e6766.png)

So, it seemed like cmake could not find where `AHardwareBuffer_describe` is defined (not declared) for Android.

I'm not familiar with NDK codes, but I think I'm missing something with that..

**Comment by byumjin** [comment]
Is there a way or can I have a chance to test `NDK21` beta version to build the latest validation layers codes?

It seems like when we use `update_external_sources_android.bat` to compile, we need to use NDK .
The problem is the current NDK20 generates obj files (related with spirV-Tool) not compatible with the current latest codes.
And, In CMake in `layerlib` folder, I've checked that I can't set the SDK target version higher than 24 with the current NDK20.
(This might cause the `undefined reference to `AHardwareBuffer_describe` error above)

**Comment by mark-lunarg** [comment]
Add gradle build to travis -- see [VulkanSamples repo.](https://github.com/LunarG/VulkanSamples/blob/master/.travis.yml)
